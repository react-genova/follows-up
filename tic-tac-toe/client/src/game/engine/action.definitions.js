export const BEGIN_GAME = 'BEGIN_GAME';
export const END_GAME = 'END_GAME';
export const EXIT_GAME = 'EXIT_GAME';
