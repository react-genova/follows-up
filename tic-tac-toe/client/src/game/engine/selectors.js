import { createSelector } from 'reselect';
import { IDLE } from './types/game.status.constants';

let getEngineRoot = state => state;

export const setEngineRoot = (getter) => {
    getEngineRoot = getter;
};

const getEngine = state => getEngineRoot(state);

const getEngineHistory = createSelector([getEngine], engine => engine.get('history'));

export const getGameStatus = createSelector([getEngine], engine => engine.get('status'));

export const getLastResult = createSelector([getEngine], engine => engine.get('result'));

export const getTotalDraws = createSelector([getEngineHistory], engine => engine.get('draws'));

export const getTotalPlayer1Victories = createSelector([getEngineHistory], engine => engine.get('player1victories'));

export const getTotalPlayer2Victories = createSelector([getEngineHistory], engine => engine.get('player2victories'));

export const isGameStarted = createSelector([getGameStatus], status => IDLE !== status);
