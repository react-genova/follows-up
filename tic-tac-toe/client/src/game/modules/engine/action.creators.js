import { BEGIN_GAME, END_GAME, EXIT_GAME } from './action.definitions';

export const beginGame = () => ({
    type: BEGIN_GAME,
    payload: { },
});

export const endGame = gameresult => ({
    type: END_GAME,
    payload: {
        gameresult,
    },
});

export const exitGame = () => ({
    type: EXIT_GAME,
    payload: { },
});
