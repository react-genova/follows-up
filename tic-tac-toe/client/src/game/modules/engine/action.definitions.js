export const BEGIN_GAME = 'BEGIN_GAME@game';
export const END_GAME = 'END_GAME@game';
export const EXIT_GAME = 'EXIT_GAME@game';
