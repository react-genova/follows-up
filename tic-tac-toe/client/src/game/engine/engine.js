import { initialEngineState } from './types/engine.types';
import { BEGIN_GAME, END_GAME, EXIT_GAME } from './action.definitions';
import {
    DRAW, PLAYER_1_WON, PLAYER_2_WON, NONE,
} from './types/game.results.constants';
import { STARTED, WAITING_FOR_START } from './types/game.status.constants';

const getFieldFromResult = gameresult => ({
    [DRAW]: 'draws',
    [PLAYER_1_WON]: 'player1victories',
    [PLAYER_2_WON]: 'player2victories',
}[gameresult]);

const engine = (state = initialEngineState, action = {}) => {
    const { type, payload } = action;
    switch (type) {
    case BEGIN_GAME: return state.withMutations((mutable) => {
        mutable.set('status', STARTED);
        mutable.set('result', NONE);
    });
    case END_GAME: return state.withMutations((mutable) => {
        const { gameresult } = payload;
        mutable.updateIn(['history', getFieldFromResult(gameresult)], value => value + 1);
        mutable.set('status', WAITING_FOR_START);
        mutable.set('result', gameresult);
    });
    case EXIT_GAME: return initialEngineState;
    default: return state;
    }
};

export default engine;
