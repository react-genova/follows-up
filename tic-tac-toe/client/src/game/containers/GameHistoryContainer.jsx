import React from 'react';
import { connect } from 'react-redux';
import GamesHistory from '../components/GamesHistory';
import {
    getPlayer1Name, getPlayer2Name, getPlayer1Symbol, getPlayer2Symbol,
} from '../../settings/modules/selectors';
import { getTotalPlayer1Victories, getTotalPlayer2Victories, getTotalDraws } from '../modules/engine/selectors';

export const GameHistoryContainer = props => <GamesHistory {...props} />;

export const mapStateToProps = state => ({
    player1name: getPlayer1Name(state),
    player2name: getPlayer2Name(state),
    player1score: getTotalPlayer1Victories(state),
    player2score: getTotalPlayer2Victories(state),
    player1sign: getPlayer1Symbol(state),
    player2sign: getPlayer2Symbol(state),
    draws: getTotalDraws(state),

});

export default connect(mapStateToProps)(GameHistoryContainer);
