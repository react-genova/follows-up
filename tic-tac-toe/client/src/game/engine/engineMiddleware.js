import {
    getPlayer1Ready, getPlayer2Ready, getPlayer1Symbol,
} from '../../settings/modules/selectors';
import { isGameStarted, isGameIdle } from './selectors';
import { beginGame, endGame } from './action.creators';
import { getBoardValues } from '../../board/modules/selectors';
import getGameResults from './engine.utils';
import { DRAW, PLAYER_1_WON, PLAYER_2_WON } from './types/game.results.constants';

const engineMiddleware = ({ getState, dispatch }) => next => (action) => {
    const resultAction = next(action);
    const state = getState();
    if (isGameStarted(state)) {
        const result = getGameResults(getBoardValues(state));
        if (result.ended) {
            if (result.draw) {
                dispatch(endGame(DRAW));
            } else if (result.winner === getPlayer1Symbol(state)) {
                dispatch(endGame(PLAYER_1_WON));
            } else {
                dispatch(endGame(PLAYER_2_WON));
            }
        }
    } else if (isGameIdle(state) && getPlayer1Ready(state) && getPlayer2Ready(state)) {
        dispatch(beginGame());
    }
    return resultAction;
};

export default engineMiddleware;
