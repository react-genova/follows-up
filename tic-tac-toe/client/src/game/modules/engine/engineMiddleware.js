import {
    getPlayer1Ready, getPlayer2Ready, getPlayer1Symbol, getAutoplay,
} from '../../../settings/modules/selectors';
import { isGameStarted, isGameIdle } from './selectors';
import { beginGame, endGame } from './action.creators';
import { getBoardResults } from '../../../board/modules/selectors';
import { DRAW, PLAYER_1_WON, PLAYER_2_WON } from './types/game.results.constants';
import { asyncDispatch } from '../_common/utils.unsafe';

const engineMiddleware = ({ getState, dispatch }) => next => (action) => {
    const resultAction = next(action);
    const state = getState();
    if (isGameStarted(state)) {
        const result = getBoardResults(state);
        if (result.ended) {
            if (result.draw) {
                dispatch(endGame(DRAW));
            } else if (result.winner === getPlayer1Symbol(state)) {
                dispatch(endGame(PLAYER_1_WON));
            } else {
                dispatch(endGame(PLAYER_2_WON));
            }
            if (getAutoplay(state)) {
                asyncDispatch(dispatch, beginGame(), 1000);
            }
        }
    } else if (isGameIdle(state) && getPlayer1Ready(state) && getPlayer2Ready(state)) {
        dispatch(beginGame());
    }
    return resultAction;
};

export default engineMiddleware;
