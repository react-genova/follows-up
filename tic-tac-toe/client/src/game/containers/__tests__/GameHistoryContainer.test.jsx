import React from 'react';
import { render, cleanup } from 'react-testing-library';
import 'jest-dom/extend-expect';
import { GameHistoryContainer, mapStateToProps } from '../GameHistoryContainer';
import { SIGN_X, SIGN_O } from '../../../board/modules/types/signs.constants';
import { getTotalDraws, getTotalPlayer1Victories, getTotalPlayer2Victories } from '../../modules/engine/selectors';
import {
    getPlayer1Symbol, getPlayer1Name, getPlayer2Name, getPlayer2Symbol,
} from '../../../settings/modules/selectors';

jest.mock('../../../settings/modules/selectors');
jest.mock('../../modules/engine/selectors');

describe('Board conatiner', () => {
    afterEach(() => {
        cleanup();
        jest.clearAllMocks();
    });

    it('renders without crashing', () => {
        const { container } = render(
            <GameHistoryContainer
                draws={1}
                player1score={1}
                player2score={1}
                player1sign={SIGN_O}
                player1name=""
                player2name=""
                player2sign={SIGN_X}
            />,
        );
        expect(container.firstChild).toMatchSnapshot();
    });

    it('creates valid props from state', () => {
        // setting up mocks
        const STATE = 'I am state';
        getTotalDraws.mockImplementation(() => 1);
        getTotalPlayer1Victories.mockImplementation(() => 2);
        getTotalPlayer2Victories.mockImplementation(() => 3);
        getPlayer1Symbol.mockImplementation(() => 4);
        getPlayer1Name.mockImplementation(() => 5);
        getPlayer2Name.mockImplementation(() => 6);
        getPlayer2Symbol.mockImplementation(() => 7);
        // invoking the method to test
        const props = mapStateToProps(STATE);
        // check result
        expect(getTotalDraws).toHaveBeenCalledTimes(1);
        expect(getTotalPlayer1Victories).toHaveBeenCalledTimes(1);
        expect(getTotalPlayer2Victories).toHaveBeenCalledTimes(1);
        expect(getPlayer1Symbol).toHaveBeenCalledTimes(1);
        expect(getPlayer1Name).toHaveBeenCalledTimes(1);
        expect(getPlayer2Name).toHaveBeenCalledTimes(1);
        expect(getPlayer2Symbol).toHaveBeenCalledTimes(1);
        expect(getTotalDraws).toHaveBeenCalledWith(STATE);
        expect(getTotalPlayer1Victories).toHaveBeenCalledWith(STATE);
        expect(getTotalPlayer2Victories).toHaveBeenCalledWith(STATE);
        expect(getPlayer1Symbol).toHaveBeenCalledWith(STATE);
        expect(getPlayer1Name).toHaveBeenCalledWith(STATE);
        expect(getPlayer2Name).toHaveBeenCalledWith(STATE);
        expect(getPlayer2Symbol).toHaveBeenCalledWith(STATE);

        expect(props).toEqual({
            draws: 1,
            player1score: 2,
            player2score: 3,
            player1sign: 4,
            player1name: 5,
            player2name: 6,
            player2sign: 7,
        });
    });
});
