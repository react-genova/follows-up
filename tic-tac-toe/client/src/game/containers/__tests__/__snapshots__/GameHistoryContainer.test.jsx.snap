// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GamesHistory conatiner renders without crashing 1`] = `
<div
  class="sc-gzVnrw hVUXwm"
>
  <span
    class="sc-ifAKCX dWYKsW"
  >
    GAME HISTORY
  </span>
  <div
    class="sc-EHOje grIMnK"
  >
    <div
      class="sc-bZQynM cQrfPv"
    >
      <div
        class="sc-bwzfXH cPrHFb"
      >
        <div
          style="width: 5em; height: 1em;"
        >
          <div
            class="sc-bdVaJa hWaRAZ"
          >
            <svg
              class="sign-o"
              height="100%"
              viewBox="0 0 100 100"
              width="100%"
            >
              <ellipse
                cx="50"
                cy="50"
                fill="transparent"
                rx="46"
                ry="46"
                stroke="yellow"
                stroke-width="8"
              />
            </svg>
          </div>
        </div>
        <span
          class="sc-htpNat jHttyq"
        >
           victories
        </span>
        <span
          class="sc-bxivhb dyMRuN"
        >
          1
        </span>
      </div>
      <div
        class="sc-bwzfXH cPrHFb"
      >
        <div
          style="width: 5em; height: 1em;"
        >
          <div
            class="sc-bdVaJa hWaRAZ"
          >
            <svg
              class="sign-x"
              height="100%"
              viewBox="0 0 100 100"
              width="100%"
            >
              <line
                stroke="cyan"
                stroke-width="8"
                x1="0"
                x2="100"
                y1="0"
                y2="100"
              />
              <line
                stroke="cyan"
                stroke-width="8"
                x1="0"
                x2="100"
                y1="100"
                y2="0"
              />
            </svg>
          </div>
        </div>
        <span
          class="sc-htpNat jHttyq"
        >
           victories
        </span>
        <span
          class="sc-bxivhb dyMRuN"
        >
          1
        </span>
      </div>
      <div
        class="sc-bwzfXH cPrHFb"
      >
        <div
          style="width: 2.5em; height: 1em;"
        >
          <div
            class="sc-bdVaJa hWaRAZ"
          >
            <svg
              class="sign-o"
              height="100%"
              viewBox="0 0 100 100"
              width="100%"
            >
              <ellipse
                cx="50"
                cy="50"
                fill="transparent"
                rx="46"
                ry="46"
                stroke="yellow"
                stroke-width="8"
              />
            </svg>
          </div>
        </div>
        <div
          style="width: 2.5em; height: 1em;"
        >
          <div
            class="sc-bdVaJa hWaRAZ"
          >
            <svg
              class="sign-x"
              height="100%"
              viewBox="0 0 100 100"
              width="100%"
            >
              <line
                stroke="cyan"
                stroke-width="8"
                x1="0"
                x2="100"
                y1="0"
                y2="100"
              />
              <line
                stroke="cyan"
                stroke-width="8"
                x1="0"
                x2="100"
                y1="100"
                y2="0"
              />
            </svg>
          </div>
        </div>
        <span
          class="sc-htpNat jHttyq"
        >
          Draws
        </span>
        <span
          class="sc-bxivhb dyMRuN"
        >
          1
        </span>
      </div>
    </div>
    <div
      class="sc-bwzfXH cMCOYb"
    >
      <span
        class="sc-bxivhb dyMRuN"
      >
        3
      </span>
    </div>
  </div>
</div>
`;
